File: simulation/api/schemas/simulation.py
"""Request/response schemas for the simulation run API."""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel, field_validator, model_validator

from feeds.algorithms.interfaces import FeedAlgorithmMetadata
from simulation.core.models.actions import TurnAction
from simulation.core.models.metrics import ComputedMetrics
from simulation.core.models.runs import RunStatus
from simulation.core.validators import (
    validate_feed_algorithm,
    validate_num_agents,
    validate_num_turns,
)


class RunRequest(BaseModel):
    """Request body for POST /v1/simulations/run."""

    num_agents: int
    num_turns: int | None = None
    feed_algorithm: str | None = None

    @field_validator("num_agents")
    @classmethod
    def _validate_num_agents(cls, v: int) -> int:
        return validate_num_agents(v)

    @field_validator("num_turns")
    @classmethod
    def _validate_num_turns(cls, v: int | None) -> int | None:
        return validate_num_turns(v)

    @field_validator("feed_algorithm")
    @classmethod
    def _validate_feed_algorithm(cls, v: str | None) -> str | None:
        return validate_feed_algorithm(v)


class FeedAlgorithmSchema(FeedAlgorithmMetadata):
    """API response for GET /v1/simulations/feed-algorithms."""

    id: str  # algorithm_id from registry


class ErrorDetail(BaseModel):
    """Error payload included when status is failed or on server error."""

    code: str
    message: str
    detail: str | None = None


class RunResponseStatus(str, Enum):
    """Status of a simulation run as returned by the API."""

    COMPLETED = "completed"
    FAILED = "failed"


class RunResponse(BaseModel):
    """Response body for POST /v1/simulations/run."""

    run_id: str
    created_at: str
    status: RunResponseStatus
    num_agents: int
    num_turns: int
    turns: list[TurnSummaryItem]
    run_metrics: ComputedMetrics | None = None
    error: ErrorDetail | None = None

    @model_validator(mode="after")
    def _validate_status_error_consistency(self) -> "RunResponse":
        if self.status == RunResponseStatus.FAILED and self.error is None:
            raise ValueError("error must be set when status is 'failed'")
        if self.status == RunResponseStatus.COMPLETED and self.error is not None:
            raise ValueError("error must be None when status is 'completed'")
        return self


class DefaultConfigSchema(BaseModel):
    """Default config for simulation start form."""

    num_agents: int
    num_turns: int


class RunConfigDetail(BaseModel):
    """Configuration for a persisted run."""

    num_agents: int
    num_turns: int
    feed_algorithm: str


class RunListItem(BaseModel):
    """Summary item for listing simulation runs."""

    run_id: str
    created_at: str
    total_turns: int
    total_agents: int
    status: RunStatus


class CreateAgentRequest(BaseModel):
    """Request body for POST /v1/simulations/agents.

    Fast-follows (not yet supported):
    - comments: list of {text, postUri}
    - likedPostUris: list of post URIs
    - linkedAgentHandles: list of agent handles to link
    """

    handle: str
    display_name: str
    bio: str = ""

    @field_validator("handle")
    @classmethod
    def _validate_handle(cls, v: str) -> str:
        from lib.validation_utils import validate_non_empty_string

        return validate_non_empty_string(v.strip(), "handle")

    @field_validator("display_name")
    @classmethod
    def _validate_display_name(cls, v: str) -> str:
        from lib.validation_utils import validate_non_empty_string

        return validate_non_empty_string(v.strip(), "display_name")


class AgentSchema(BaseModel):
    """Agent profile for the simulation UI."""

    handle: str
    name: str
    bio: str
    generated_bio: str
    followers: int
    following: int
    posts_count: int


class FeedSchema(BaseModel):
    """Feed metadata for one agent in a turn."""

    feed_id: str
    run_id: str
    turn_number: int
    agent_handle: str
    post_uris: list[str]
    created_at: str


class PostSchema(BaseModel):
    """Post content for display in agent feeds. Matches ApiPost in ui/lib/api/simulation.ts."""

    uri: str
    author_display_name: str
    author_handle: str
    text: str
    bookmark_count: int
    like_count: int
    quote_count: int
    reply_count: int
    repost_count: int
    created_at: str


class AgentActionSchema(BaseModel):
    """Action event performed by an agent in a turn."""

    action_id: str
    agent_handle: str
    post_uri: str | None = None
    user_id: str | None = None
    type: TurnAction
    created_at: str


class TurnSchema(BaseModel):
    """Full turn payload consumed by the UI."""

    turn_number: int
    agent_feeds: dict[str, FeedSchema]
    agent_actions: dict[str, list[AgentActionSchema]]


class TurnActionsItem(BaseModel):
    """One turn summary with aggregate action counts."""

    turn_number: int
    created_at: str
    total_actions: dict[str, int]
    metrics: ComputedMetrics | None = None


class RunDetailsResponse(BaseModel):
    """Response body for GET /v1/simulations/runs/{run_id}."""

    run_id: str
    status: RunStatus
    created_at: str
    started_at: str
    completed_at: str | None = None
    config: RunConfigDetail
    turns: list[TurnActionsItem]
    run_metrics: ComputedMetrics | None = None


class TurnSummaryItem(BaseModel):
    """One turn summary with computed metrics."""

    turn_number: int
    created_at: str
    total_actions: dict[str, int]
    metrics: ComputedMetrics

File: ui/lib/api/simulation.ts
import {
  Agent,
  AgentAction,
  ApiError,
  Feed,
  FeedAlgorithm,
  Post,
  Run,
  RunConfig,
  Turn,
} from '@/types';
import { FALLBACK_DEFAULT_CONFIG } from '@/lib/default-config';
import type { components } from '@/types/api.generated';

const DEFAULT_SIMULATION_API_BASE_URL: string = 'http://localhost:8000/v1';
const SIMULATION_API_BASE_URL: string = (
  process.env.NEXT_PUBLIC_SIMULATION_API_URL || DEFAULT_SIMULATION_API_BASE_URL
).replace(/\/$/, '');

type ApiAgent = components['schemas']['AgentSchema'];

function mapAgent(apiAgent: ApiAgent): Agent {
  return {
    handle: apiAgent.handle,
    name: apiAgent.name,
    bio: apiAgent.bio,
    generatedBio: apiAgent.generated_bio,
    followers: apiAgent.followers,
    following: apiAgent.following,
    postsCount: apiAgent.posts_count,
  };
}

type ApiRunListItem = components['schemas']['RunListItem'];

type ApiFeed = components['schemas']['FeedSchema'];

type ApiAgentAction = components['schemas']['AgentActionSchema'];

type ApiDefaultConfig = components['schemas']['DefaultConfigSchema'];

type ApiFeedAlgorithm = components['schemas']['FeedAlgorithmSchema'];

type ApiRunResponse = components['schemas']['RunResponse'];

type ApiPost = components['schemas']['PostSchema'];

type ApiTurn = components['schemas']['TurnSchema'];

function buildApiUrl(path: string): string {
  return `${SIMULATION_API_BASE_URL}${path}`;
}

/** Token getter for authenticated requests. Set by AuthProvider. */
let authTokenGetter: (() => Promise<string | null>) | null = null;

export function setAuthTokenGetter(
  getter: (() => Promise<string | null>) | null,
): void {
  authTokenGetter = getter;
}

/** Called when a request returns 401. Set by AuthProvider to trigger signOut. */
let onUnauthorized: (() => void) | null = null;

export function setOnUnauthorized(callback: (() => void) | null): void {
  onUnauthorized = callback;
}

async function getAuthHeaders(): Promise<Record<string, string>> {
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };
  const token = authTokenGetter ? await authTokenGetter() : null;
  if (token != null && token !== '') {
    headers['Authorization'] = `Bearer ${token}`;
  }
  return headers;
}

/**
 * Handles non-ok response: parses error payload and throws ApiError.
 */
async function handleErrorResponse(response: Response): Promise<never> {
  if (response.status === 401 && onUnauthorized) {
    onUnauthorized();
  }
  const responseText: string = await response.text();
  let code: string = 'UNKNOWN_ERROR';
  let message: string = responseText || `Request failed (${response.status})`;
  let detail: string | null = null;

  try {
    const data: unknown = JSON.parse(responseText);
    const err =
      data && typeof data === 'object' && data !== null && 'error' in data
        ? (data as { error?: { code?: string; message?: string; detail?: string | null } }).error
        : null;
    if (err && typeof err === 'object' && err !== null) {
      code = (typeof err.code === 'string' && err.code) || code;
      message = (typeof err.message === 'string' && err.message) || message;
      detail = typeof err.detail === 'string' ? err.detail : err.detail === null ? null : null;
    }
  } catch {
    // Non-JSON body: use fallback
  }

  throw new ApiError(code, message, detail, response.status);
}

/**
 * Fetches JSON from a URL with GET. Throws ApiError on non-2xx responses.
 * Includes Authorization Bearer token when authTokenGetter is set.
 */
async function fetchJson<T>(url: string): Promise<T> {
  const headers = await getAuthHeaders();
  const response: Response = await fetch(url, {
    method: 'GET',
    headers,
  });

  if (!response.ok) {
    await handleErrorResponse(response);
  }

  return response.json() as Promise<T>;
}

/**
 * POSTs JSON to a URL. Throws ApiError on non-2xx responses.
 */
async function fetchPost<TReq, TRes>(url: string, body: TReq): Promise<TRes> {
  const headers = await getAuthHeaders();
  const response: Response = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    await handleErrorResponse(response);
  }

  return response.json() as Promise<TRes>;
}

function mapFeed(apiFeed: ApiFeed): Feed {
  return {
    feedId: apiFeed.feed_id,
    runId: apiFeed.run_id,
    turnNumber: apiFeed.turn_number,
    agentHandle: apiFeed.agent_handle,
    postUris: apiFeed.post_uris,
    createdAt: apiFeed.created_at,
  };
}

function mapAction(apiAction: ApiAgentAction): AgentAction {
  return {
    actionId: apiAction.action_id,
    agentHandle: apiAction.agent_handle,
    postUri: apiAction.post_uri ?? undefined,
    userId: apiAction.user_id ?? undefined,
    type: apiAction.type,
    createdAt: apiAction.created_at,
  };
}

function mapPost(apiPost: ApiPost): Post {
  return {
    uri: apiPost.uri,
    authorDisplayName: apiPost.author_display_name,
    authorHandle: apiPost.author_handle,
    text: apiPost.text,
    bookmarkCount: apiPost.bookmark_count,
    likeCount: apiPost.like_count,
    quoteCount: apiPost.quote_count,
    replyCount: apiPost.reply_count,
    repostCount: apiPost.repost_count,
    createdAt: apiPost.created_at,
  };
}

function mapTurn(apiTurn: ApiTurn): Turn {
  const agentFeeds: Record<string, Feed> = {};
  const agentActions: Record<string, AgentAction[]> = {};

  Object.entries(apiTurn.agent_feeds).forEach(([agentHandle, apiFeed]) => {
    agentFeeds[agentHandle] = mapFeed(apiFeed);
  });

  Object.entries(apiTurn.agent_actions).forEach(([agentHandle, apiActions]) => {
    agentActions[agentHandle] = apiActions.map(mapAction);
  });

  return {
    turnNumber: apiTurn.turn_number,
    agentFeeds,
    agentActions,
  };
}

export async function getDefaultConfig(): Promise<RunConfig> {
  const api: ApiDefaultConfig = await fetchJson<ApiDefaultConfig>(
    buildApiUrl('/simulations/config/default'),
  );
  return {
    numAgents: api.num_agents,
    numTurns: api.num_turns,
    feedAlgorithm: FALLBACK_DEFAULT_CONFIG.feedAlgorithm,
    feedAlgorithmConfig: null,
  };
}

function mapFeedAlgorithm(api: ApiFeedAlgorithm): FeedAlgorithm {
  return {
    id: api.id,
    displayName: api.display_name,
    description: api.description,
    configSchema: api.config_schema ?? null,
  };
}

export async function getFeedAlgorithms(): Promise<FeedAlgorithm[]> {
  const api: ApiFeedAlgorithm[] = await fetchJson<ApiFeedAlgorithm[]>(
    buildApiUrl('/simulations/feed-algorithms'),
  );
  return api.map(mapFeedAlgorithm);
}

export async function postRun(config: RunConfig): Promise<Run> {
  const body = {
    num_agents: config.numAgents,
    num_turns: config.numTurns,
    feed_algorithm: config.feedAlgorithm,
    feed_algorithm_config: config.feedAlgorithmConfig,
  };
  const api: ApiRunResponse = await fetchPost<typeof body, ApiRunResponse>(
    buildApiUrl('/simulations/run'),
    body,
  );
  return {
    runId: api.run_id,
    createdAt: api.created_at,
    totalTurns: api.num_turns,
    totalAgents: api.num_agents,
    status: api.status,
  };
}

export async function getRuns(): Promise<Run[]> {
  const apiRuns: ApiRunListItem[] = await fetchJson<ApiRunListItem[]>(
    buildApiUrl('/simulations/runs'),
  );

  return apiRuns.map((run) => ({
    runId: run.run_id,
    createdAt: run.created_at,
    totalTurns: run.total_turns,
    totalAgents: run.total_agents,
    status: run.status,
  }));
}

export async function getTurnsForRun(runId: string): Promise<Record<string, Turn>> {
  const apiTurnsById: Record<string, ApiTurn> = await fetchJson<Record<string, ApiTurn>>(
    buildApiUrl(`/simulations/runs/${encodeURIComponent(runId)}/turns`),
  );

  const turnsById: Record<string, Turn> = {};
  Object.entries(apiTurnsById).forEach(([turnId, apiTurn]) => {
    turnsById[turnId] = mapTurn(apiTurn);
  });

  return turnsById;
}

export async function getAgents(params?: {
  limit?: number;
  offset?: number;
}): Promise<Agent[]> {
  const limit: number | undefined = params?.limit;
  const offset: number | undefined = params?.offset;

  const baseUrl: string = buildApiUrl('/simulations/agents');
  const parts: string[] = [];
  if (limit != null) {
    parts.push(`limit=${encodeURIComponent(String(limit))}`);
  }
  if (offset != null) {
    parts.push(`offset=${encodeURIComponent(String(offset))}`);
  }
  const url: string = parts.length > 0 ? `${baseUrl}?${parts.join('&')}` : baseUrl;

  const apiAgents: ApiAgent[] = await fetchJson<ApiAgent[]>(url);
  return apiAgents.map(mapAgent);
}

/** Returns mock agents for run-detail context (dummy runs). */
export async function getMockAgents(): Promise<Agent[]> {
  const apiAgents: ApiAgent[] = await fetchJson<ApiAgent[]>(
    buildApiUrl('/simulations/agents/mock'),
  );
  return apiAgents.map(mapAgent);
}

export async function postAgent(body: {
  handle: string;
  display_name: string;
  bio?: string;
}): Promise<Agent> {
  const api: ApiAgent = await fetchPost<typeof body, ApiAgent>(
    buildApiUrl('/simulations/agents'),
    body,
  );
  return mapAgent(api);
}

export async function getPosts(uris?: string[]): Promise<Post[]> {
  const baseUrl: string = buildApiUrl('/simulations/posts');
  const url: string =
    uris != null && uris.length > 0
      ? `${baseUrl}?${uris.map((u) => `uris=${encodeURIComponent(u)}`).join('&')}`
      : baseUrl;
  const apiPosts: ApiPost[] = await fetchJson<ApiPost[]>(url);
  return apiPosts.map(mapPost);
}
